crates/bevy_pbr/src/meshlet/from_mesh.rs

```diff
@@ -19,80 +26,343 @@ impl MeshletMesh {
+                meshlet.vertex_count as usize,
+            );
+        }
+    }
+
+    meshlets
+}
+
+fn collect_triangle_edges_per_meshlet(
+    simplification_queue: Range<usize>,
+    meshlets: &Meshlets,
+) -> HashMap<usize, HashSet<(u32, u32)>> {
+    let mut triangle_edges_per_meshlet = HashMap::new();
+    for meshlet_id in simplification_queue {
+        let meshlet = meshlets.get(meshlet_id);
+        let meshlet_triangle_edges = triangle_edges_per_meshlet
+            .entry(meshlet_id)
+            .or_insert(HashSet::new());
+        for i in meshlet.triangles.chunks(3) {
+            let v0 = meshlet.vertices[i[0] as usize];
+            let v1 = meshlet.vertices[i[1] as usize];
+            let v2 = meshlet.vertices[i[2] as usize];
+            meshlet_triangle_edges.insert((v0.min(v1), v0.max(v1)));
+            meshlet_triangle_edges.insert((v0.min(v2), v0.max(v2)));
+            meshlet_triangle_edges.insert((v1.min(v2), v1.max(v2)));
+        }
+    }
+    triangle_edges_per_meshlet
+}
+
+fn find_connected_meshlets(
+    simplification_queue: Range<usize>,
+    triangle_edges_per_meshlet: &HashMap<usize, HashSet<(u32, u32)>>,
+) -> HashMap<usize, Vec<(usize, usize)>> {
+    let mut connected_meshlets_per_meshlet = HashMap::new();
+    for meshlet_id in simplification_queue.clone() {
+        connected_meshlets_per_meshlet.insert(meshlet_id, Vec::new());
+    }
+
+    for (meshlet_id1, meshlet_id2) in simplification_queue.tuple_combinations() {
+        let shared_edge_count = triangle_edges_per_meshlet[&meshlet_id1]
+            .intersection(&triangle_edges_per_meshlet[&meshlet_id2])
+            .count();
+        if shared_edge_count != 0 {
+            connected_meshlets_per_meshlet
+                .get_mut(&meshlet_id1)
+                .unwrap()
+                .push((meshlet_id2, shared_edge_count));
+            connected_meshlets_per_meshlet
+                .get_mut(&meshlet_id2)
+                .unwrap()
+                .push((meshlet_id1, shared_edge_count));
+        }
+    }
+    connected_meshlets_per_meshlet
+}
+
+fn group_meshlets(
+    simplification_queue: Range<usize>,
+    connected_meshlets_per_meshlet: &HashMap<usize, Vec<(usize, usize)>>,
+) -> HashMap<i32, Vec<usize>> {
+    let mut xadj = Vec::with_capacity(simplification_queue.len() + 1);
+    let mut adjncy = Vec::new();
+    let mut adjwgt = Vec::new();
+    for meshlet_id in simplification_queue.clone() {
+        xadj.push(adjncy.len() as i32);
+        for (connected_meshlet_id, shared_edge_count) in
+            connected_meshlets_per_meshlet[&meshlet_id].iter().copied()
+        {
+            adjncy.push((connected_meshlet_id - simplification_queue.start) as i32);
+            adjwgt.push(shared_edge_count as i32);
+        }
+    }
+    xadj.push(adjncy.len() as i32);
+
+    let mut group_per_meshlet = vec![0; simplification_queue.len()];
+    let partition_count = (simplification_queue.len().div_ceil(4)) as i32;
```

Comment:
Why 4?

Reply:
4 meshlets per group, ideally. 4 meshlets (256 triangles), simplify group to 50% (128 triangles), split back into meshlets = 2 meshlets (64 x 2 triangles). 

I should probably rename this to target_partition_count, there are no hard limits.

Reply:
Earlier code mentions that the meshlets are grouped into _roughly_ 4 in a group. How come 4 is always used here if there aren't guaranteed to be 4?

Reply:
It's the partition count. Given N nodes (meshlets), Metis divides them into P (partition_count) partitions (groups).

So given 1000 meshlets, we tell Metis to divide the 1000 meshlets into (1000/25)=250 groups. It's not guaranteed that we have _exactly_ 4 per group though. Some groups might have slightly more or less than 4 meshlets, as it tries to spread the cost out as defined by the edge weights between nodes. We just have to hope for an even split between all partitions, which would give 4 meshlets per group if that were to happen.